docs/rfcs/2025-07-23-global-gc-worker.md

```diff
@@ -0,0 +1,113 @@
+
+
+#### Handling Unused Files
+
+With the integration of the GC worker into the major compaction process, the risk of accidentally deleting newly created SST files that have not yet been recorded in the manifest is significantly mitigated. Consequently, the concept of "Unused Files" as a distinct category primarily susceptible to accidental deletion is largely resolved. Any files that are genuinely "unused" (i.e., never referenced by any manifest, including temporary ones) can be safely deleted after a configurable maximum lingering time.
+
+For debugging and auditing purposes, a comprehensive list of recently deleted files can be maintained.
+
+### Ensuring Read Consistency
+
+To prevent the GC worker from inadvertently deleting files that are actively being utilized by long-running analytical queries, particularly on read replicas, a robust protection mechanism is introduced. Upon detection of a long-running query (e.g., by the dedicated slow query recorder after a query have been running for a while), a temporary manifest will be written to the region's manifest directory. This temporary manifest will explicitly reference all files necessary for the ongoing query, along with the relevant manifest version. This action effectively places a hold on these files, preventing their deletion by the GC worker. The lifecycle of this temporary manifest, including its creation and eventual deletion upon query completion or cancellation, will be managed by the slow query recorder.
+
+## Drawbacks
+
+- **Dependency on Compaction Frequency**: The integration of the GC worker with major compaction means that GC cycles are directly tied to the frequency of compactions. In environments with infrequent compaction operations, obsolete files may accumulate for extended periods before being reclaimed, potentially leading to increased storage consumption.
+- **Race Condition with Long-Running Queries**: A potential race condition exists if a long-running query initiates but haven't write its temporary manifest in time, while a compaction process simultaneously begins and marks files used by that query as obsolete. This scenario could lead to the premature deletion of files still required by the active query. To mitigate this, the threshold time for writing a temporary manifest should be significantly shorter than the lingering time configured for obsolete files, ensuring that next GC worker runs do not delete files that are now referenced by a temporary manifest if the query is still running.
+Also the read replica shouldn't be later in manifest version for more than the lingering time of obsolete files, otherwise it might ref to files that are already deleted by the GC worker.
+- need to upload tmp manifest to object storage, which may introduce additional complexity and potential performance overhead. But since long-running queries are typically rare, the performance impact is expected to be minimal.
```

Comment:
In historical analytics, long-running queries may be common.

Reply:
> In historical analytics, long-running queries may be common.

Oh by "rare" here I mean so that uploading a tmp manifest is a acceptable cost compare to having to wait a long time and had the query failed